### PR TITLE
fix: 修复大量数据下，行列总计单元格计算时耗时和内存问题

### LINKED ISSUE #1516: 🐛 大数据量下，同时开启，行列总计，卡死
### 🏷 Version

| Package      | Version |
| -------------- | --------- |
| @antv/s2       |       1.22.0   |
| @antv/s2-react |         1.20.0|
| @antv/s2-vue   |          |

### Sheet Type
PivotSheet

- [-] PivotSheet
- [ ] TableSheet
- [ ] GridAnalysisSheet
- [ ] StrategySheet

### 🖋 Description
大数据量下，交叉表，同时开启行、列总计，页面卡死，单独开某一个正常

### ⌨️ Code Snapshots
https://codesandbox.io/s/empty-framework-j55inp?file=/src/App.tsx

### PR DESCRIPTION
**Pull request by wjgogogo**
### 👀 PR includes

✨ Feature

- [ ] New feature

🎨 Enhance

- [ ] Code style optimization
- [ ] Refactoring
- [ ] Change the UI
- [ ] Improve the performance
- [ ] Type optimization

🐛 Bugfix

- [x] Solve the issue and close #1516 

🔧 Chore

- [ ] Test case
- [ ] Docs / demos update
- [ ] CI / workflow
- [ ] Release version
- [ ] Other ()

### 📝 Description
之前的`flatten`计算对于内存和性能消耗都很大，如果表格数据量特别大时，并且遇到单元格刚好是行总计和列总计的交叉单元格，s2会将所有的数据拿到然后再进行`flatten`，会直接卡死
### 🖼️ Screenshot
在64w条数据的情况下，`flatten`性能消耗：
| Before | After |
| ------ | ----- |
| ❌   卡死   |  image |

### 🔍 Self-Check before the merge

- [ ] Add or update relevant docs.
- [ ] Add or update relevant demos.
- [ ] Add or update test case.
- [ ] Add or update relevant TypeScript definitions.

### CODE REVIEW

File: packages/s2-core/src/utils/data-set-operate.ts [hunk 1]
```diff
@@ -44,21 +44,19 @@ export const flattenDeep = (data: Record<any, any>[] | Record<any, any>) =>
...
-      const currentKeys = keys(current);
-      if (currentKeys.includes('undefined')) {
-        currentKeys.forEach((ki) => {
+    data.forEach((current) => {
+      if (current && 'undefined' in current) {
+        keys(current).forEach((ki) => {
           result.push(current[ki]);
         });
       } else {
-        result = result.concat(current);
```
Comment (wjgogogo) [inline packages/s2-core/src/utils/data-set-operate.ts:57]: `concat`会频繁触发gc，对内存和性能都是挑战

**Review by lijinke666 — DISMISSED** [review]

**Review by lijinke666 — APPROVED** [review]

**Review by xingwanying — COMMENTED** [review]
加个单测？

**Comment by wjgogogo** [comment]
@xingwanying 这部分已经有单测覆盖了

**Review by lijinke666 — APPROVED** [review]

**Pull request merged**